node/node_cross_shard.go

```diff
@@ -120,124 +120,77 @@ func (node *Node) VerifyBlockCrossLinks(block *types.Block) error {
 		)
 	}
 
-	firstCrossLinkBlock := core.EpochFirstBlock(node.Blockchain().Config().CrossLinkEpoch)
-
-	for i, crossLink := range *crossLinks {
-		lastLink := &types.CrossLink{}
-		if i == 0 {
-			if crossLink.BlockNum().Cmp(firstCrossLinkBlock) > 0 {
-				lastLink, err = node.Blockchain().ReadShardLastCrossLink(crossLink.ShardID())
-				if err != nil {
-					return ctxerror.New("[CrossLinkVerification] no last cross link found 1",
-						"blockHash", block.Hash(),
-						"crossLink", lastLink,
-					).WithCause(err)
-				}
-			}
-		} else {
-			if (*crossLinks)[i-1].Header().ShardID() != crossLink.Header().ShardID() {
-				if crossLink.BlockNum().Cmp(firstCrossLinkBlock) > 0 {
-					lastLink, err = node.Blockchain().ReadShardLastCrossLink(crossLink.ShardID())
-					if err != nil {
-						return ctxerror.New("[CrossLinkVerification] no last cross link found 2",
-							"blockHash", block.Hash(),
-							"crossLink", lastLink,
-						).WithCause(err)
-					}
-				}
-			} else {
-				lastLink = &(*crossLinks)[i-1]
-			}
-		}
-
-		if crossLink.BlockNum().Cmp(firstCrossLinkBlock) > 0 { // TODO: verify genesis block
-			err = node.VerifyCrosslinkHeader(lastLink.Header(), crossLink.Header())
-			if err != nil {
-				return ctxerror.New("cannot ValidateNewBlock",
+	for _, crossLink := range *crossLinks {
+		cl, err := node.Blockchain().ReadCrossLink(crossLink.ShardID(), crossLink.BlockNum())
+		if err == nil && cl != nil {
+			if !bytes.Equal(cl.Serialize(), crossLink.Serialize()) {
+				return ctxerror.New("[CrossLinkVerification] Double signed crossLink",
 					"blockHash", block.Hash(),
-					"numTx", len(block.Transactions()),
-				).WithCause(err)
+					"Previous committed crossLink", cl,
+					"crossLink", crossLink,
+				)
 			}
+			continue
+		}
+		if err = node.VerifyCrossLink(crossLink); err != nil {
+			return ctxerror.New("cannot VerifyBlockCrossLinks",
+				"blockHash", block.Hash(),
+				"blockNum", block.Number(),
+				"crossLinkShard", crossLink.ShardID(),
+				"crossLinkBlock", crossLink.BlockNum(),
+				"numTx", len(block.Transactions()),
+			).WithCause(err)
 		}
 	}
 	return nil
 }
 
-// ProcessHeaderMessage verify and process Node/Header message into crosslink when it's valid
-func (node *Node) ProcessHeaderMessage(msgPayload []byte) {
-	if node.NodeConfig.ShardID == 0 {
-
-		var headers []*block.Header
-		err := rlp.DecodeBytes(msgPayload, &headers)
+// ProcessCrossLinkMessage verify and process Node/CrossLink message into crosslink when it's valid
+func (node *Node) ProcessCrossLinkMessage(msgPayload []byte) {
+	// TODO: non-leader in beaconchain doesn't need process crosslink message, but still need to monitor leader's behavior
+	if node.NodeConfig.ShardID == 0 && node.Consensus.IsLeader() {
```

Comment:
Everyone needs to process cross-links locally and remove them once it's confirmed in blocks. Otherwise, if the leader is down or leader switched, the new leader won't have the pending crosslinks to propose, then those links will be lost.

Reply:
I move the delete pending crosslinks to insertchain code.